classes/core/PKPString.php

```diff
@@ -392,6 +397,127 @@ public static function getWordCount(string $str): int
+            'y' => 'yy',   // A two digit representation of a year
+            // Time
+            'a' => 'a',    // Lowercase Ante meridiem and Post meridiem, am or pm
+            'A' => 'a',    // Uppercase Ante meridiem and Post meridiem, AM or PM
+            'B' => '',     // Swatch Internet time, no ICU equivalent
+            'g' => 'h',    // 12-hour format of an hour without leading zeros
+            'G' => 'H',    // 24-hour format of an hour without leading zeros
+            'h' => 'hh',   // 12-hour format of an hour with leading zeros
+            'H' => 'HH',   // 24-hour format of an hour with leading zeros
+            'i' => 'mm',   // Minutes with leading zeros
+            's' => 'ss',   // Seconds with leading zeros
+            'v' => '',     // Milliseconds, no direct ICU equivalent
+            // Timezone
+            'e' => 'zzzz', // Timezone identifier
+            'I' => '',     // Whether or not the date is in daylight saving time, no ICU equivalent
+            'O' => 'xx',   // Difference to Greenwich time (GMT) without colon between hours and minutes
+            'P' => 'xxx',  // Difference to Greenwich time (GMT) with colon between hours and minutes
+            'T' => 'z',    // Timezone abbreviation
+            'Z' => 'X',    // Timezone offset in seconds
+            // Full Date/Time
+            'c' => 'yyyy-MM-dd\'T\'HH:mm:ssXXX', // ISO 8601 date
+            'r' => 'r',    // RFC 2822 formatted date
+            'U' => 'U',    // Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)
+        ];
+
+        $icuFormat = strtr($phpFormat, $mapping);
+        
+        return $icuFormat;
+    }
+
+    /**
+     * Add @calendar suffix to format for special cases
+     */
+    private static function getLocaleWithCalendar($locale) 
+    {
+        // Mapping of locales to their default calendar systems
+        $localeCalendars = [
+            'th' => '@calendar=buddhist', // Thai
+            'zh_CN' => '@calendar=chinese', // Chinese (Simplified)
+            'zh_TW' => '@calendar=roc', // Taiwanese
+            'am' => '@calendar=ethiopic', // Amharic
+            'ar_SA' => '@calendar=islamic-umalqura', // Arabic (Saudi Arabia)
+            'he' => '@calendar=hebrew', // Hebrew
+            'fa' => '@calendar=persian', // Persian
+            'ja_JP' => '@calendar=japanese', // Japanese
+        ];
+
+        if (array_key_exists($locale, $localeCalendars)) {
+            $locale = $locale . $localeCalendars[$locale];
+        }
+
+        return $locale;
+    }
+
+    /**
+     * Convert any datetime object to a localized string taking into account the 
+     * datetime format.
+     * 
+     * @throws \Exception General exception for potential errors during the date format
+     */
+    public static function getLocalizedDate(string|int|DateTimeInterface|null $date, string $format, ?string $locale = null): ?string
```

Comment:
@defstat, I'd consider this PR as-is, but part of our trouble with dates is that we throw around a lot of different kinds of dates without distinguishing clearly between them -- numeric timestamps and string representations, mainly. A side-effect of this is that I'm never sure what I'm dealing with, and any date processing ends up with a mess of conditionally called `strtotime` or `Carbon::parse` calls. Now that we're adopting typehints and able to use the `IntlDateFormatter`, I think we should standardize on `DateTimeImmutable`, and write functions that expect them consistently as inputs. Conversion to/from other formats should move close to the database layer (for stored values) and to the handler layer (for user inputs) and everything in between can use properly typed data. What do you think?